### PR TITLE
Add R web application linking with OpenCPU server

### PR DESCRIPTION
**Pull request by niio972**
Add R web app stand alone integration.
Modifications
-  add opensilex/opencpu-client-php": "dev-master to composer
- add data-analysis ( controller, views, model)

### CODE REVIEW

File: composer.lock [hunk 1]
```diff
@@ -3186,23 +3296,26 @@
         },
         {
             "name": "codeception/phpunit-wrapper",
-            "version": "7.7.1",
+            "version": "6.6.1",
```
Comment (Gouga34) [inline composer.lock]: Can you explain why you have changed all the composer.lock ? Especially the update of dependencies version ?
For example here, why have you downgraded the codeception/phpunit-wrapper required version ?

File: controllers/DataAnalysisController.php [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+                ]
+            ]
+        ];
+    }
+
+    /**
+     * List all R apps
+     * @return mixed
+     */
+    public function actionIndex($integrated = false) {
```
Comment (Gouga34) [inline controllers/DataAnalysisController.php]: The documentation for the parameter $integrated is missing

File: controllers/DataAnalysisController.php [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+     */
+    public function actionView() {
+        $searchParams = Yii::$app->request->queryParams;
+        return $this->render('iframe-view', [
+                    'appUrl' => $searchParams["url"],
+                    'appName' => $searchParams["name"]
+                        ]
+        );
+    }
+
```
Comment (Gouga34) [inline controllers/DataAnalysisController.php]:
```suggestion
```

File: models/yiiModels/DataAnalysisAppSearch.php [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+ */
+class DataAnalysisAppSearch {
+    /**
+     *
+     * @var OpenCPUServer 
+     */
+    public $ocpuserver;
+
+    /**
+     * Configuration file constants
```
Comment (Gouga34) [inline models/yiiModels/DataAnalysisAppSearch.php]: Can you explain which configuration file is it ? You could maybe add its path ?

File: controllers/DataAnalysisController.php [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+/**
+ * Implements the controller for available statistical and visualisation programs
+ * @see yii\web\Controller
+ * @author Arnaud Charleroy <arnaud.charleroy@inra.fr>
+ */
+class DataAnalysisController extends \yii\web\Controller {
+    /**
+     * Constants used to parse configuration file
+     */
```
Comment (Gouga34) [inline controllers/DataAnalysisController.php]: Can you explain which configuration file is it ? You could maybe add its path ?

File: config/config.php [hunk 6]
```diff
@@ -31,6 +31,7 @@ public static function path() {
             'basePath' => '/'.$basePath.'/',
             'hostnameURL' => 'http://'.$hostname.'/'.$appli,
             'documentsUrl' => '../web/documents/',
+            'ocpuServer' => 'http://opensilex.org:8004/ocpu/',
```
Comment (Gouga34) [inline config/config.php]:
```suggestion
            'ocpuServer' => 'http://localhost:8004/ocpu/',
```

File: models/yiiModels/DataAnalysisAppSearch.php [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+            }
+        }
+        return $appList;
+    }
+    
+    /**
+     * List all available apps
+     * @param array $sessionToken used for the data access
+     * @param string $params search params
+     * @return array list of app with their informations
```
Comment (Gouga34) [inline models/yiiModels/DataAnalysisAppSearch.php]: Add an example of the returned array.

File: models/yiiModels/DataAnalysisAppSearch.php [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+        foreach ($apps as $app) {
+
+            $appConfiguration = $this->getAppConfiguration($app);
+
+            if (!empty($appConfiguration) 
+                    && isset($appConfiguration[self::AVAILABLE_FUNCTIONS]) 
+                    && isset($appConfiguration[self::INTEGRATED_FUNCTION])) {
+                $availableFunctions = $appConfiguration[self::AVAILABLE_FUNCTIONS];
+                $integratedFunctions = $appConfiguration[self::INTEGRATED_FUNCTION];
+                foreach ($availableFunctions as $functionName) {
```
Comment (Gouga34) [inline models/yiiModels/DataAnalysisAppSearch.php]: Add some documentation inside the method by commenting the steps ;)

File: models/yiiModels/DataAnalysisAppSearch.php [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+                    $visualisationsInfo[$functionName][self::APP_INDEX_HREF] = Url::to(['data-analysis/view', 'url' => $url, 'name' => explode("/", $app)[1]]);
+                }
+            }
+        }
+        return $visualisationsInfo;
+    }
+
+    /**
+     * Retreive information on demo app only
+     * @return array information on demo app
```
Comment (Gouga34) [inline models/yiiModels/DataAnalysisAppSearch.php]: Add @example of the returned array form

File: models/yiiModels/DataAnalysisAppSearch.php [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+        $visualisationsInfo = [];
+
+        if ($this->ocpuserver->status()) {
+            $app = self::DEFAULT_DEMO_APP;
+
+            $appConfiguration = $this->getAppConfiguration($app);
+            if (!empty($appConfiguration) 
+                    && isset($appConfiguration[self::AVAILABLE_FUNCTIONS][0]) 
+                    && isset($appConfiguration[self::INTEGRATED_FUNCTION])) {
+                // only one function
```
Comment (Gouga34) [inline models/yiiModels/DataAnalysisAppSearch.php]: Could you be more precised on this comment ?

File: models/yiiModels/DataAnalysisAppSearch.php [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+                $visualisationsInfo[self::APP_INDEX_HREF] = Url::to(['data-analysis/view', 'url' => $url, 'name' => explode("/", $app)[1]]);
+            }
+        }
+        return $visualisationsInfo;
+    }
+
+    /**
+     * Return exported functionalities 
+     * @param string $applicationName 
+     * @return array functionalities
```
Comment (Gouga34) [inline models/yiiModels/DataAnalysisAppSearch.php]: Add @example of the format of the returned array

File: models/yiiModels/DataAnalysisAppSearch.php [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+            }
+        }
+        return [];
+    }
+
+  
+     /**
+     * Return R application configuration
+     * @param string $app
+     * @return array
```
Comment (Gouga34) [inline models/yiiModels/DataAnalysisAppSearch.php]: Add an @example of the format of the returned array

File: views/data-analysis/error.php [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+// Contact: arnaud.charleroy@inra.fr, anne.tireau@inra.fr, pascal.neveu@inra.fr
+//******************************************************************************
+use yii\helpers\Html;
+use kartik\icons\Icon;
+
+/* @var $this yii\web\View */
+
+$this->title = Yii::t('app', '{n, plural, =1{Stat/Vizu Application} other{Stat/Vizu Applications}}', ['n' => 2]);
+$this->params['breadcrumbs'][] = $this->title;
+if(!isset($message)){
```
Comment (Gouga34) [inline views/data-analysis/error.php]:
```suggestion
if (!isset($message)) {
```

File: views/data-analysis/error.php [hunk 14]
```diff
@@ -0,0 +1,25 @@
+<?php
+
+//******************************************************************************
```
Comment (Gouga34) [inline views/data-analysis/error.php]: Why don't you use the /site/error page ? 

(in your controller : 

```php
return $this->render('/site/error', [
                    'name' => Yii::t('app/messages','Internal error'),
                    'message' => Yii::t('app/messages', 'An error occurred.')]);
```
)

File: views/data-analysis/index.php [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+                    "data-analysis/run-script/", 
+                    "function" => $function,
+                    "rpackage" => $appInfo[DataAnalysisAppSearch::R_PACKAGE_NAME]
+                    ]);
+    } else {
+        $appHref = $appInfo[DataAnalysisAppSearch::APP_INDEX_HREF];
+    }
+    echo Html::beginTag("div", ["class" => "col-sm-6 col-md-5"]);
+    echo Html::beginTag("div", ["class" => "thumbnail"]);
+    echo Html::beginTag("a", ["href" => $appHref]);
```
Comment (Gouga34) [inline views/data-analysis/index.php]: Why don't you use the Html::img method ? 
Maybe something such as this example : 
`return Html::a(Html::img(Yii::getAlias('@img-url'). '/myimage.jpg'),['index'])`

File: views/layouts/main.php [hunk 16]
```diff
@@ -143,6 +143,18 @@
 //                            ]
                             //\SILEX:info
                         ]];
+        $menuItems[] = ['label' => Yii::t('app', 'Data Analysis'),
```
Comment (Gouga34) [inline views/layouts/main.php]: Maybe you should add a config parameter for the data analysis part. If a platform don't have this module, they should not have the link in the menu.

File: views/layouts/main.php [hunk 17]
```diff
@@ -143,6 +143,18 @@
 //                            ]
                             //\SILEX:info
                         ]];
```
Comment (Gouga34) [inline views/layouts/main.php]: Only the admin can have the Data Analysis tab in the menu. Is this normal?

File: controllers/DataAnalysisController.php [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+     * @return type
+     */
+    public function actionViewDemo() {
+        $searchModel = new DataAnalysisAppSearch();
+        $appDemoInformation = $searchModel->getAppDemoInformation();
+        if (!empty($appDemoInformation)) {
+            $this->redirect($appDemoInformation[DataAnalysisAppSearch::APP_INDEX_HREF]);
+        } else {
+            return $this->render('error', [
+                        'message' => 'Demo application not available.'
```
Comment (annetireau) [inline controllers/DataAnalysisController.php]: Use translation system

File: models/yiiModels/DataAnalysisAppSearch.php [hunk 19]
```diff
@@ -0,0 +1,232 @@
...
+    const APP_SHORT_NAME = "appShortName";
+    const R_PACKAGE_NAME = "packageName";
+    const APP_INDEX_HREF = "appIndexHref";
+    const FUNCTION_NAME = "functionName";
+    const FUNCTION_HELP = "functionHelp";
+    const AVAILABLE_FUNCTIONS = "availableFunctions";
+    const FUNCTION_DESCRIPTION = "description";
+    const INTEGRATED_FUNCTION = "integratedFunctions";
+    
+    const DEFAULT_DEMO_APP = "opensilex/opensilex-datavis-rapp-demo";
```
Comment (annetireau) [inline models/yiiModels/DataAnalysisAppSearch.php]: Could you explain all this const ?

File: models/yiiModels/DataAnalysisAppSearch.php (hunk 7 above)
Comment (annetireau) [reply models/yiiModels/DataAnalysisAppSearch.php]: Add more explaination of the process and link between app and opencpu, add example

File: models/yiiModels/DataAnalysisAppSearch.php (hunk 11 above)
Comment (annetireau) [reply models/yiiModels/DataAnalysisAppSearch.php]: Add explanation

File: models/yiiModels/DataAnalysisAppSearch.php (hunk 12 above)
Comment (annetireau) [reply models/yiiModels/DataAnalysisAppSearch.php]: Explain link/use with yaml

File: models/yiiModels/DataAnalysisAppSearch.php [hunk 20]
```diff
@@ -0,0 +1,232 @@
...
+  
+     /**
+     * Return R application configuration
+     * @param string $app
+     * @return array
+     */
+    function getAppConfiguration($app) {
+        if ($this->ocpuserver->status()) {
+            try {
+                $response = $this->ocpuserver->getOpenCPUWebServerClient()->request(OpenCPUServer::OPENCPU_SERVER_GET_METHOD, 'apps/' . $app . '/opensilex/webAppConfig.yml');
```
Comment (annetireau) [inline models/yiiModels/DataAnalysisAppSearch.php]: Add explanation

File: views/data-analysis/error.php [hunk 21]
```diff
@@ -0,0 +1,25 @@
...
+use yii\helpers\Html;
+use kartik\icons\Icon;
+
+/* @var $this yii\web\View */
+
+$this->title = Yii::t('app', '{n, plural, =1{Stat/Vizu Application} other{Stat/Vizu Applications}}', ['n' => 2]);
+$this->params['breadcrumbs'][] = $this->title;
+if(!isset($message)){
+    $message = 'An error has occurred';
```
Comment (annetireau) [inline views/data-analysis/error.php]: Use translation process

File: views/data-analysis/iframe-view.php [hunk 22]
```diff
@@ -0,0 +1,28 @@
...
+use yii\helpers\Html;
+use kartik\icons\Icon;
+
+/* @var $this yii\web\View */
+
+$this->title = Yii::t('app', '{n, plural, =1{Stat/Vizu Application} other{Stat/Vizu Applications}}', ['n' => 2]);
+$this->params['breadcrumbs'][] = $this->title;
+echo Html::beginTag('div', ['class' => 'row']);
+echo Html::tag('h5',
+        Icon::show('flask', ['class' => 'fa-large'], Icon::FA). "Experimental version",
```
Comment (annetireau) [inline views/data-analysis/iframe-view.php]: Add translation
Show it only if there are apps

File: views/layouts/main.php (hunk 17 above)
Comment (annetireau) [reply views/layouts/main.php]: no

File: views/layouts/main.php (hunk 16 above)
Comment (annetireau) [reply views/layouts/main.php]: good idea

File: views/data-analysis/index.php (hunk 15 above)
Comment (niio972) [reply views/data-analysis/index.php]: I added it.

File: views/layouts/main.php (hunk 17 above)
Comment (niio972) [reply views/layouts/main.php]: Corrected .

File: views/data-analysis/iframe-view.php (hunk 22 above)
Comment (niio972) [reply views/data-analysis/iframe-view.php]: Done

File: views/data-analysis/error.php (hunk 14 above)
Comment (niio972) [reply views/data-analysis/error.php]: Done

File: views/layouts/main.php (hunk 16 above)
Comment (niio972) [reply views/layouts/main.php]: Nice I add it.

File: models/yiiModels/DataAnalysisAppSearch.php (hunk 19 above)
Comment (niio972) [reply models/yiiModels/DataAnalysisAppSearch.php]: Done

File: composer.lock (hunk 1 above)
Comment (niio972) [reply composer.lock]: I update the composer in the wrong way. The commit c6647cf0f68203874be71030727ed932ae8384da correct this issue.

File: controllers/DataAnalysisController.php (hunk 2 above)
Comment (niio972) [reply controllers/DataAnalysisController.php]: Done.

File: models/yiiModels/DataAnalysisAppSearch.php (hunk 4 above)
Comment (niio972) [reply models/yiiModels/DataAnalysisAppSearch.php]: The configuration file is use to integrate the form with Yii2 it has been remove from this pull request.

File: controllers/DataAnalysisController.php (hunk 18 above)
Comment (niio972) [reply controllers/DataAnalysisController.php]: Done !

File: models/yiiModels/DataAnalysisAppSearch.php (hunk 20 above)
Comment (niio972) [reply models/yiiModels/DataAnalysisAppSearch.php]: I add some application R structure comments to this function.

File: views/data-analysis/error.php (hunk 21 above)
Comment (niio972) [reply views/data-analysis/error.php]: Done.

File: models/yiiModels/DataAnalysisAppSearch.php (hunk 4 above)
Comment (Gouga34) [reply models/yiiModels/DataAnalysisAppSearch.php]: Explain the configuration file content in the code, not just for me ;)

**Pull request merged**